File: 0490-the-maze/0490-the-maze.py
class Solution:
    def hasPath(self, maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
        if maze is None or len(maze) == 0:
            return False

        m = len(maze)
        n = len(maze[0])

        queue = collections.deque()
        queue.append(start)
        maze[start[0]][start[1]] = 2
        dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        while queue:
            curr = queue.popleft()
            
            for d in dirs:
                i = curr[0]
                j = curr[1]
                while (i < m and j < n and i >= 0 and j >= 0 and maze[i][j] != 1):
                    i += d[0]
                    j += d[1]
                
                i -= d[0]
                j -= d[1]

                if(maze[i][j] == 0):
                    if i == destination[0] and j == destination[1]:
                        return True
                    maze[i][j] = 2
                    queue.append([i, j])
        
        return False
        

File: 0490-the-maze/0490-the-maze.java
class Solution {
    int[][]dirs;
    int m;
    int n;
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        if(maze == null || maze.length == 0){
            return false;
        }
         m = maze.length;
         n = maze[0].length;
        dirs = new int[][]{{0, 1},{0, -1},{1, 0},{-1, 0}};

        return dfs(maze, start, destination);
    }

    private boolean dfs(int[][]maze, int[] start, int[] destination){
        //base
        if(maze[start[0]][start[1]] == 2){
            return false;
        }

        //logic
        if(start[0] == destination[0] && start[1] == destination[1]){
            return true;
        }
        maze[start[0]][start[1]] = 2;
        //visited
        for(int[] dir: dirs){
            int i = start[0];
            int j = start[1];
            while(i < m && j < n && i >=0 && j >= 0 && maze[i][j] != 1){
                i += dir[0];
                j += dir[1];
            }
            i -= dir[0];
            j -= dir[1];

            if(dfs(maze, new int[]{i, j}, destination)) return true;
        }
        return false;
    }
}


//DFS
//TIme Complexity: O(m*n)
//Space Complexity:O(m⋅n)